File: design/erlang/day1/fizzbuzzwhizz.erl
-module(fizzbuzzwhizz).
-compile(export_all).

times(N) ->
  fun(M) ->
    M rem N =:= 0
  end.

contains(N) ->
  fun(M) ->
    P1 = M rem 10,
    P2 = M div 10 rem 10,
    P3 = M div 10 div 10 rem 10,
    P1 =:= N orelse P2 =:= N orelse P3 =:= N
  end.

rule(Pred, Trans) ->
  fun(N) ->
    case Pred(N) of
      true  -> {true, Trans(N)};
      false -> false
    end
  end.

'AND'(R5) ->
  'AND'(R5, fun(A1, A2) -> A1++A2 end, "").

'AND'([], _Op, Acc) -> fun(_N) -> {true, Acc} end;
'AND'([R|Rest], Op, Acc) ->
  fun(N) ->
    case R(N) of
      {true,A} -> ('AND'(Rest, Op, Op(Acc, A)))(N);
      false    -> false
    end
  end.

'OR'(R5) ->
  'OR'(R5, fun(A1, A2) -> A1++A2 end, "").

'OR'([], _Op, Acc) -> fun(_N) -> false end;
'OR'([R|Rest], Op, Acc) ->
  fun(N) ->
    case R(N) of
      {true,A} -> {true, Op(Acc, A)};
      false    -> ('OR'(Rest, Op, Acc))(N)
    end
  end.


test1() ->
  R1_3 = rule(times(3), fun(_) -> "Fizz" end),
  R1_5 = rule(times(5), fun(_) -> "Buzz" end),
  R1_7 = rule(times(7), fun(_) -> "Whizz" end),
  
  R1 = 'OR'([R1_3, R1_5, R1_7]),
  
  R2 = 'OR'(['AND'([R1_3, R1_5, R1_7]),
             'AND'([R1_3, R1_5]),
             'AND'([R1_3, R1_7]),
             'AND'([R1_5, R1_7])]),
  
  R3 = rule(contains(3), fun(_) -> "Fizz" end),
  
  Rd = rule(fun(_) -> true end, fun(N) -> integer_to_list(N) end),
  
  Spec = 'OR'([R3, R2, R1, Rd]),
  
  [Spec(I) || I<-lists:seq(1,20)].

File: design/scala/src/main/scala/day1/FizzBuzzWhizz.scala
package day1

object FizzBuzzWhizz {
	type Rule = Int => Option[String]
	type Op = (String, String) => String
	def times(n: Int) = (x: Int) => x % n == 0
	def contains(n: Int) = (x: Int) => x.toString contains n.toString
	def rule(pred: Int => Boolean, trans: Int => String): Rule = (x: Int) => if (pred(x)) Some(trans(x)) else None
	def OR(rules: List[Rule]): Rule = rules match {
	  case r :: rest => (x: Int) => r(x) match {
	    case Some(a) => Some(a)
	    case None => OR(rest)(x)
	  }
	  case Nil => (x: Int) => None
	}
	def AND(rules: List[Rule], op: Op = (acc, a) => acc + a, acc: String = ""): Rule = rules match {
	  case r :: rest => (x: Int) => r(x) match {
	    case Some(a) => AND(rest, op, op(acc, a))(x)
	    case None => None
	  }
	  case Nil => (x: Int) => Some(acc)
	}
	
	// foldLeft implementation
	def OR1(rules: List[Rule]): Rule = 
	  (x: Int) => rules.foldLeft(None: Option[String])((acc, r) => acc match {
	  	case Some(a) => Some(a)
	  	case None => r(x)
	  })
	def AND1(rules: List[Rule], op: Op = (acc, a) => acc + a): Rule = 
	  (x: Int) => rules.foldLeft(Some(""): Option[String])((acc, r) => acc match {
	  	case Some(a) => r(x) match {
	  	  case Some(b) => Some(op(a, b))
	  	  case None => None
	  	}
	  	case None => None
	  })
	
	val R1_3 = rule(times(3), (_) => "Fizz")
	val R1_5 = rule(times(5), (_) => "Buzz")
	val R1_7 = rule(times(7), (_) => "Whizz")
	val R1 = OR1(List(R1_3, R1_5, R1_7))
	val R2 = OR1(List(AND1(List(R1_3, R1_5, R1_7)),
	                 AND1(List(R1_3, R1_5)),
	                 AND1(List(R1_5, R1_7))))
	val R3 = rule(contains(3), (_) => "Fizz")
	val Rd = rule((_) => true, _.toString)
	val Spec = OR1(List(R3, R2, R1, Rd))
	
	def main(args: Array[String]): Unit = {
	  1 to 20 map Spec map (_.get) foreach println
	}
}